src/React/Hook.purs

```diff
@@ -0,0 +1,310 @@
+module React.Hook
+  ( Hook
+
+  , useState
+  , useStateLazy
+  , setState
+  , modifyState
+  , SetState
+
+  , useEffect
+  , effectInput
+  , EffectInput
+
+  , useContext
+  , Context
+
+  , useReducer
+  , useReducerLazy
+  , dispatch
+  , Dispatch
+
+  , useCallback
+  , callbackInput
+  , CallbackInput
+
+  , useMemo
+  , memoInput
+  , MemoInput
+
+  , useRef
+  , getRef
+  , setRef
+  , Ref
+
+  , useImperativeMethods
+  , imperativeMethodsInput
+  , ImperativeMethodsInput
+
+  , useMutationEffect
+  , mutationEffectInput
+  , MutationEffectInput
+
+  , useLayoutEffect
+  , layoutEffectInput
+  , LayoutEffectInput
+  ) where
+
+import Prelude
+
+import Data.Function.Uncurried (Fn1, Fn2, Fn3, Fn4, mkFn2, runFn1, runFn2, runFn3, runFn4)
+import Data.Maybe (Maybe)
+import Data.Nullable (Nullable)
+import Data.Nullable as Nullable
+import Data.Tuple (Tuple(..))
+
+import Effect (Effect)
+import Effect.Uncurried (EffectFn1, EffectFn2, runEffectFn1, runEffectFn2)
+
+import Unsafe.Coerce (unsafeCoerce)
+
+useState
+  :: forall a
+   . a
+  -> Hook (Tuple a (SetState a))
```

Comment:
I'd be interested to see if the state handling internally is fully parametric or if it still has all the sugar around copying/merging properties that component `setState` uses.

Reply:
I think one reason why this can't be parametric is that the `setState` state function is overloaded to either take the state as is or a function, which means you can't have a function as state.

Reply:
I believe this is the relevant section handling the state:
https://github.com/facebook/react/blob/52bea95cfce4a75bc7c1a09455abaeef50fa0700/packages/react-reconciler/src/ReactFiberHooks.js#L333-L472

The only thing I see is that the code checks if the state is a function:
https://github.com/facebook/react/blob/52bea95cfce4a75bc7c1a09455abaeef50fa0700/packages/react-reconciler/src/ReactFiberHooks.js#L455-L457

As you mention it seems the overloading is problematic. I am open to ideas or suggests for the representation on this.

Reply:
Unfortunately there's no safe way to describe the API. One option would be to enumerate JS primitives. This is conservative.

```purescript
foreign import data StateValue :: Type -> Type

intValue :: Int -> StateValue Int
intValue = unsafeCoerce

stringValue :: String -> StateValue String
stringValue = unsafeCoerce

recordValue :: forall r. { | r } -> StateValue { | r }
recordValue = unsafeCoerce

useState :: forall a. StateValue a -> Hook (Tuple a (SetState a))
```

This could potentially be overloaded with an instance chain that closes it off with a custom type error. You could then expose an `unsafeUseState` function which documents the caveat that you can't use anything which has a function as it's representation.